src/library_fs.js

```diff
@@ -1018,10 +1048,32 @@ FS.staticInit();
     },
     ftruncate(fd, len) {
       var stream = FS.getStreamChecked(fd);
+      var node = stream.node;
       if ((stream.flags & {{{ cDefs.O_ACCMODE }}}) === {{{ cDefs.O_RDONLY}}}) {
         throw new FS.ErrnoError({{{ cDefs.EINVAL }}});
       }
-      FS.truncate(stream.node, len);
+      if (!node.node_ops.setattr && !stream.stream_ops.setattr) {
+        throw new FS.ErrnoError({{{ cDefs.EPERM }}});
+      }
+      if (FS.isDir(node.mode)) {
+        throw new FS.ErrnoError({{{ cDefs.EISDIR }}});
+      }
+      if (!FS.isFile(node.mode)) {
+        throw new FS.ErrnoError({{{ cDefs.EINVAL }}});
+      }
+      var errCode = FS.nodePermissions(node, 'w');
+      if (errCode) {
+        throw new FS.ErrnoError(errCode);
+      }
+      var attrs = {
+        size: len,
+        timestamp: Date.now()
+      };
+      if (stream.stream_ops.setattr) {
+        stream.stream_ops.setattr(stream, attrs);
+      } else {
+        node.node_ops.setattr(node, attrs);
```

Comment:
Is this a common pattern where we try to do an operation first on the stream and then on the node if it fails?

Do we need the fallback?

Reply:
Well we have a map from file descriptors to streams, and a map from paths to nodes. Each stream has a reference to its node. The difficulty is that not every node has a valid path. 

Most file systems store all relevant information about a node on the node itself. However, in the nodefs, each stream has a distinct native file descriptor. In the case where the node has no valid path, the only way to talk to the native object is with `fstat` and friends using the native file descriptor. So in that case we need these `stream_ops`. For every other FS, it's sufficient to have `setattr` on the `node_ops`.

Also, we _do_ need the `node_ops` case because if we call `stat` instead of `fstat` then we have a path but no file descriptor. We could `open` the path to get the file descriptor and then `stat` it and close it. But we're better off just having a separate handler for gettattr/setattr on a path/node than on a file descriptor/stream. But only nodefs seems to need it.